src/lib/coda_base/data_hash.ml

```diff
@@ -146,13 +137,30 @@ struct
   [%%endif]
 end
 
-module Make_full_size () = struct
+module Make_full_size (B58_data : Data_hash_intf.Data_hash_descriptor) = struct
   module Basic = Make_basic (struct
     let length_in_bits = Field.size_in_bits
   end)
 
   include Basic
 
+  module Base58_check = Codable.Make_base58_check(struct
+    type nonrec t = t
+
+    include Hashable.Make_binable(Basic)
+
+    include B58_data
+  end)
+
+  [%%define_locally
+  Base58_check.(to_base58_check, of_base58_check, of_base58_check_exn)]
+
+  [%%define_locally
+  Base58_check.String_ops.(to_string, of_string)]
+
+  [%%define_locally
+  Base58_check.(to_yojson, of_yojson)]
+
   (* inside functor of no arguments, versioned types are allowed *)
```

Comment:
@psteckler this comment is what I'm now violating - should this be structured differently? Do you have some advice?

Reply:
@emberian The solution we've used is to have an unversioned type inside the functor, and create a versioned type at each place the functor is applied. 

An example is in `Consensus.Proof_of_stake`, the module `Staking_value_versioned`.